airbyte-webapp/src/components/connection/CatalogTree/streamConfigHelpers/streamConfigHelpers.ts

```diff
@@ -0,0 +1,96 @@
+  config: AirbyteStreamConfiguration,
+  selectedCursorField: string[],
+  numberOfFieldsInStream: number
+): Partial<AirbyteStreamConfiguration> => {
+  const previouslySelectedFields = config?.selectedFields || [];
+
+  // If field selection is enabled, we need to be sure the new cursor is also selected
+  if (config?.fieldSelectionEnabled && !previouslySelectedFields.find((field) => isEqual(field, selectedCursorField))) {
+    if (previouslySelectedFields.length === numberOfFieldsInStream - 1) {
+      return { cursorField: selectedCursorField, selectedFields: [], fieldSelectionEnabled: false };
+    }
+    return {
+      fieldSelectionEnabled: true,
+      selectedFields: [...previouslySelectedFields, { fieldPath: selectedCursorField }],
+      cursorField: selectedCursorField,
+    };
+  }
+  return { cursorField: selectedCursorField };
+};
+
+/**
+ * Overwrites the entire primaryKey value in AirbyteStreamConfiguration.
+ */
+export const updatePrimaryKey = (
+  config: AirbyteStreamConfiguration,
+  compositePrimaryKey: string[][],
+  numberOfFieldsInStream: number
+): Partial<AirbyteStreamConfiguration> => {
+  const previouslySelectedFields = config?.selectedFields || [];
+
+  // If field selection is enabled, we need to be sure each fieldPath in the new primary key is also selected
+  if (
+    config?.fieldSelectionEnabled &&
+    !compositePrimaryKey.some((fieldPath) => previouslySelectedFields.find((field) => isEqual(field, fieldPath)))
+  ) {
+    if (previouslySelectedFields.length === numberOfFieldsInStream - 1) {
+      return { primaryKey: compositePrimaryKey, selectedFields: [], fieldSelectionEnabled: false };
+    }
+    return {
+      fieldSelectionEnabled: true,
+      selectedFields: [...previouslySelectedFields, ...compositePrimaryKey.map((fieldPath) => ({ fieldPath }))],
+      primaryKey: compositePrimaryKey,
+    };
+  }
+
+  return {
+    primaryKey: compositePrimaryKey,
+  };
+};
+
+/**
+ * Adds a single fieldPath to the composite primaryKey
+ */
+export const addFieldToPrimaryKey = (
+  config: AirbyteStreamConfiguration,
+  fieldPath: string[],
+  numberOfFieldsInStream: number
+): Partial<AirbyteStreamConfiguration> => {
+  const fieldIsSelected = !config?.primaryKey?.find((pk) => isEqual(pk, fieldPath));
+  const previouslySelectedFields = config?.selectedFields || [];
+  let newPrimaryKey: string[][];
+
+  if (!fieldIsSelected) {
+    newPrimaryKey = config.primaryKey?.filter((key) => !isEqual(key, fieldPath)) ?? [];
+  } else {
+    newPrimaryKey = [...(config?.primaryKey ?? []), fieldPath];
+  }
+
+  // If field selection is enabled, we need to be sure the new fieldPath is also selected
+  if (
+    fieldIsSelected &&
+    config?.fieldSelectionEnabled &&
+    !previouslySelectedFields.find((field) => isEqual(field, fieldPath))
+  ) {
+    if (previouslySelectedFields.length === numberOfFieldsInStream - 1) {
+      return { primaryKey: newPrimaryKey, selectedFields: [], fieldSelectionEnabled: false };
```

Comment:
Can you help me understand why we set `fieldSelectionEnabled: false` in this case?

Reply:
In this case the only unselected field is the field being added to the primary key. So changing `fieldSelectionEnabled: false, selectedFields: []` effectively turns off column selection (because now all fields are selected).

Your comment did make me realize there is a different bug in `updatePrimaryKey()` in this same logic, so thanks for that 🙂

Reply:
Cool!  One more question here -- it doesn't look like `fieldSelectionEnabled` is checked at (or above) the `FieldRow` level unless there was a piece I missed.  Could that help us avoid some potentially redundant calculation there?

Reply:
I don't think there's any added calculation overhead related to field selection until the callbacks that change `primaryKey` and `cursorField` which are contained in `streamConfigHelpers.ts`. Did you spot something in particular?

Reply:
`FieldRow` is evaluating if a field is a cursor or a primary key every time.  Couldn't we avoid that if we know `fieldSelectionEnabled = false` for that entire stream?

Reply:
`isCursor` and `isPrimaryKey` aren't just for field selection, though. They're also needed to know whether the cursor field radio button or primary key checkbox is checked.

Reply:
Ah, yep.  Thanks.